server/src/main/java/org/apache/druid/server/coordinator/duty/MarkEternityTombstonesAsUnused.java

```diff
@@ -0,0 +1,173 @@
+   * <li> For each such candidate segment that is a tombstone with an infinite start or end, look at the set of overshadowed
+   * segments to see if any of the intervals overlaps with the candidate segment.
+   * <li> If there is no overlap, add the candidate segment to the eternity segments result set. </li>
+   * There can at most be two such candidate tombstones per datasource  -- one that starts at {@link DateTimes#MIN}
+   * and another that ends at {@link DateTimes#MAX}. </li>
+   * </p>
+   *
+   * @param dataSourcesSnapshot the datasources snapshot for segments timeline
+   * @return the set of candidate tombstones grouped by datasource
+   */
+  private Map<String, Set<SegmentId>> determineCandidateTombstones(final DataSourcesSnapshot dataSourcesSnapshot)
+  {
+    final Map<String, Set<SegmentId>> datasourceToDanglingTombstones = new HashMap<>();
+
+    dataSourcesSnapshot.getDataSourcesMap().keySet().forEach((datasource) -> {
+      final SegmentTimeline usedSegmentsTimeline
+          = dataSourcesSnapshot.getUsedSegmentsTimelinesPerDataSource().get(datasource);
+
+      final Optional<Set<DataSegment>> usedNonOvershadowedSegments =
+          Optional.fromNullable(usedSegmentsTimeline)
+                  .transform(timeline -> timeline.findNonOvershadowedObjectsInInterval(
+                      Intervals.ETERNITY,
+                      Partitions.ONLY_COMPLETE
+                  ));
+
+      if (usedNonOvershadowedSegments.isPresent()) {
+        usedNonOvershadowedSegments.get().forEach(candidateSegment -> {
+          if (isTombstoneWithInfiniteStartOrEnd(candidateSegment)) {
+            boolean overlaps = dataSourcesSnapshot.getOvershadowedSegments().stream()
+                                                  .filter(overshadowedSegment ->
+                                                              candidateSegment.getDataSource()
+                                                                              .equals(overshadowedSegment.getDataSource()))
+                                                  .anyMatch(
+                                                      overshadowedSegment ->
+                                                          candidateSegment.getInterval()
+                                                                          .overlaps(overshadowedSegment.getInterval())
+                                                  );
+            if (!overlaps) {
+              datasourceToDanglingTombstones
+                  .computeIfAbsent(datasource, ds -> new HashSet<>())
+                  .add(candidateSegment.getId());
+            }
+          }
+        });
+      }
+    });
+
+    return datasourceToDanglingTombstones;
+  }
+
+  private boolean isTombstoneWithInfiniteStartOrEnd(final DataSegment segment)
+  {
+    return segment.isTombstone() && segment.getShardSpec().getNumCorePartitions() == 0 && (
```

Comment:
Could you please add a comment explaining why the numCorePartitions must be 0 as it is not obvious from the method name

Reply:
Or perhaps you could rename the method to something like isNewGenerationEternityTombstone as the reason is already in the javadocs.